---
title: "17_otu_clustering"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE}
library(DECIPHER)
library(Biostrings)
library(speedyseq)
```


```{r}
fung_physeq_test <- readRDS("../intermediates/raw_fungal_physeq.rds")

test_fasta <- readDNAStringSet("../pipeline_outputs/OTUs.100.fasta")
```

```{r}


d <- DECIPHER::DistanceMatrix(test_fasta, processors = 5)

clusters <- DECIPHER::IdClusters(
  d, 
  method = "complete",
  cutoff = 0.03, # corresponds to 97% OTUs
  processors = 5
)

ps0 <- merge_taxa_vec(
  fung_physeq_test, 
  group = clusters$cluster,
  tax_adjust = 2
) # 36842 OTUs vs 52112 raw ASVs
```

```{r}
# remove ASVs that are only in one sample
fung_abun <- as.data.frame(otu_table(ps0))
fung_meta <- as.data.frame(sample_data(ps0))
fung_meta <- data.frame(fung_meta)
# sample names in abundance table eventually get "X" appended due to the colnames being numeric
# make a new column with these names
fung_meta$x_seq_ids <- paste0("X", fung_meta$sequencing_id)



fung_asv_sums <- data.frame(rowSums(fung_abun))
fung_asv_sums$sample_freq <- apply(fung_abun, 1, function(c)sum(c!=0))
names(fung_asv_sums) <- c("asv_abundance", "sample_freq")

fung_asv_sums <- arrange(fung_asv_sums, desc(sample_freq))

fung_asvs_no_singlesamps <- fung_asv_sums[fung_asv_sums$sample_freq>1,]

fung_abun_no_singles <- abun[rownames(abun) %in% rownames(fung_asvs_no_singlesamps),] # 22406 ASVs remain 
fung_abun_no_singles <- fung_abun_no_singles[,colSums(fung_abun_no_singles)>0] # 1642 samples
fung_met_no_singles <- fung_meta[as.character(fung_meta$sequencing_id) %in% names(fung_abun_no_singles),]
```


```{r}
habs <- c("Terrestrial","Riverine", "Marine")
fung_meta <- fung_meta[which(fung_meta$habitat %in% habs),]

hab_subset <- function(asv_table, meta_table) {
  habs <- unique(meta_table$habitat)
  
  hab_list <- list()
  
  for (a_hab in habs) {
    sub_met <- meta_table[which(meta_table$habitat==a_hab),]
    
    sub_hab_abun <- asv_table[,which(names(asv_table) %in% sub_met$sequencing_id)]
    sub_hab_abun <- data.frame(sub_hab_abun)
    sub_hab_abun <- sub_hab_abun[which(rowSums(sub_hab_abun) > 0),]
    sub_hab_abun <- data.frame(sub_hab_abun)
    
    hab_list[[a_hab]] <- sub_hab_abun
    
  }
  
  return(hab_list)
}

fung_asvs_by_hab <- hab_subset(abun, fung_meta)



make_emp_hab_list <- function(hab, hab_list, meta_table, asv_table) {
  
  dat <- hab_list[[hab]]
  samples <- names(dat)
  
  sub_met <- meta_table[which(meta_table$x_seq_ids %in% samples),]
  
  empo3 <- unique(sub_met$empo_3)
  emp_list <- list()
  
  for (e in empo3) {
    sub_sub_met <- sub_met[which(sub_met$empo_3==e),]
    sub_abun <- asv_table[,which(names(asv_table) %in% sub_sub_met$sequencing_id)]
    sub_abun <- data.frame(sub_abun)
    sub_abun <- sub_abun[which(rowSums(sub_abun) > 0),]
    sub_abun <- data.frame(sub_abun)
    
    asv_sums <- data.frame(rowSums(sub_abun))
    asv_sums$sample_freq <- apply(sub_abun, 1, function(c)sum(c!=0))
    names(asv_sums) <- c("asv_abundance", "sample_freq")
    asv_sums$tot_samps <- ncol(sub_abun)
    asv_sums$prev <- asv_sums$sample_freq/ncol(sub_abun)
    
    asv_sums <- arrange(asv_sums, desc(sample_freq))
    
    asv_sums <- asv_sums[asv_sums$prev>0.2,]
    
    emp_list[[e]] <- asv_sums
  }
  
  return(emp_list)
}

#sites <- names(fung_asvs_by_site)

fung_emp_asv_sums_by_hab <- lapply(habs,
                            make_emp_hab_list,
                            fung_asvs_by_hab,
                            fung_meta,
                            abun)

names(fung_emp_asv_sums_by_hab) <- names(fung_asvs_by_hab)


all_things_listed <- lapply(fung_emp_asv_sums_by_hab, function(x) lapply(x, rownames))
flattened_ish <- lapply(all_things_listed, function(x) Reduce(c, x))

all_things <- unlist(flattened_ish)

all_things_unique <- unique(all_things)


fung_abun_prevfilter_on_empo <- fung_abun[rownames(fung_abun) %in% all_things_unique,]
fung_abun_prevfilter_on_empo <- fung_abun_prevfilter_on_empo[,colSums(fung_abun_prevfilter_on_empo)>0] #2051 ASVs, 1654 samples
```

## Bacteria

```{r}
# bacteria test
#otu <- fread("../16s_data/abundance_table_97.shared")
tax <- fread("../16s_data/annotations_97.taxonomy")

abun <- fread(file = "../16s_data/abundance_table_97.shared",
             drop = c("numOtus", "label", "Group"),
             header = T)

abun_names <- fread(file = "../16s_data/abundance_table_97.shared",
                   select = "Group",
                   colClasses = "character",
                   header = T)

# no idea what this does - data.table syntax?
abun <- abun[, Group := abun_names]

# move group col to front
abun <- abun %>% select(Group, everything())

abun <- column_to_rownames(abun, "Group")

# convert character cols to numeric
#abun <- as.data.frame(sapply(abun,as.numeric))

# transpose so ASVs are rows, samples are cols
abun <- as.data.frame(t(abun))

# 112283 OTUs
```

```{r}
# remove ASVs that are only in one sample
bact_abun <- abun
bact_meta <- fread("../16s_data/cleaned/mm_16s_hiseqs_metadata_table.csv")
# sample names in abundance table eventually get "X" appended due to the colnames being numeric
# make a new column with these names
bact_meta$x_seq_ids <- paste0("X", bact_meta$sequencing_id)



bact_asv_sums <- data.frame(rowSums(bact_abun))
bact_asv_sums$sample_freq <- apply(bact_abun, 1, function(c)sum(c!=0))
names(bact_asv_sums) <- c("asv_abundance", "sample_freq")

bact_asv_sums <- arrange(bact_asv_sums, desc(sample_freq))

bact_asvs_no_singlesamps <- bact_asv_sums[bact_asv_sums$sample_freq>1,]

bact_abun_no_singles <- bact_abun[rownames(bact_abun) %in% rownames(bact_asvs_no_singlesamps),] # 71458 ASVs remain 
bact_abun_no_singles <- bact_abun_no_singles[,colSums(bact_abun_no_singles)>0] # 1642 samples
bact_met_no_singles <- bact_meta[as.character(bact_meta$sequencing_id) %in% names(bact_abun_no_singles),]
```


```{r}
habs <- c("Terrestrial","Riverine", "Marine")
bact_meta <- bact_meta[which(bact_meta$habitat %in% habs),]

hab_subset <- function(asv_table, meta_table) {
  habs <- unique(meta_table$habitat)
  
  hab_list <- list()
  
  for (a_hab in habs) {
    sub_met <- meta_table[which(meta_table$habitat==a_hab),]
    
    sub_hab_abun <- asv_table[,which(names(asv_table) %in% sub_met$sequencing_id)]
    sub_hab_abun <- data.frame(sub_hab_abun)
    sub_hab_abun <- sub_hab_abun[which(rowSums(sub_hab_abun) > 0),]
    sub_hab_abun <- data.frame(sub_hab_abun)
    
    hab_list[[a_hab]] <- sub_hab_abun
    
  }
  
  return(hab_list)
}

bact_asvs_by_hab <- hab_subset(bact_abun, bact_meta)



make_emp_hab_list <- function(hab, hab_list, meta_table, asv_table) {
  
  dat <- hab_list[[hab]]
  samples <- names(dat)
  
  sub_met <- meta_table[which(meta_table$x_seq_ids %in% samples),]
  
  empo3 <- unique(sub_met$empo_3)
  emp_list <- list()
  
  for (e in empo3) {
    sub_sub_met <- sub_met[which(sub_met$empo_3==e),]
    sub_abun <- asv_table[,which(names(asv_table) %in% sub_sub_met$sequencing_id)]
    sub_abun <- data.frame(sub_abun)
    sub_abun <- sub_abun[which(rowSums(sub_abun) > 0),]
    sub_abun <- data.frame(sub_abun)
    
    asv_sums <- data.frame(rowSums(sub_abun))
    asv_sums$sample_freq <- apply(sub_abun, 1, function(c)sum(c!=0))
    names(asv_sums) <- c("asv_abundance", "sample_freq")
    asv_sums$tot_samps <- ncol(sub_abun)
    asv_sums$prev <- asv_sums$sample_freq/ncol(sub_abun)
    
    asv_sums <- arrange(asv_sums, desc(sample_freq))
    
    asv_sums <- asv_sums[asv_sums$prev>0.2,]
    
    emp_list[[e]] <- asv_sums
  }
  
  return(emp_list)
}

#sites <- names(bact_asvs_by_site)

bact_emp_asv_sums_by_hab <- lapply(habs,
                            make_emp_hab_list,
                            bact_asvs_by_hab,
                            bact_meta,
                            bact_abun)

names(bact_emp_asv_sums_by_hab) <- names(bact_asvs_by_hab)


all_things_listed <- lapply(bact_emp_asv_sums_by_hab, function(x) lapply(x, rownames))
flattened_ish <- lapply(all_things_listed, function(x) Reduce(c, x))

all_things <- unlist(flattened_ish)

all_things_unique <- unique(all_things)


bact_abun_prevfilter_on_empo <- bact_abun[rownames(bact_abun) %in% all_things_unique,]
bact_abun_prevfilter_on_empo <- bact_abun_prevfilter_on_empo[,colSums(bact_abun_prevfilter_on_empo)>0] #11858 ASVs, 1654 samples
```







