---
title: "04_prevalence_culling"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE}
library(dplyr)
library(tibble)
```

## Fungi

```{r}

fung_abun <- readRDS("../intermediates/global/fung_culled_abun_otu_lauras_method_post_matchup.rds") # 25385 otus, 1424 samples
fung_meta <- readRDS("../intermediates/global/fung_culled_metadata_otu_lauras_method_post_matchup.rds")

# remove otus that are only in one sample

# sample names in abundance table eventually get "X" appended due to the colnames being numeric
# make a new column with these names
# fung_meta$x_seq_ids <- paste0("X", fung_meta$sequencing_id)


fung_otu_sums <- data.frame(rowSums(fung_abun))
rownames(fung_otu_sums) <- rownames(fung_abun)
fung_otu_sums$sample_freq <- apply(fung_abun, 1, function(c)sum(c!=0))
names(fung_otu_sums) <- c("otu_abundance", "sample_freq")

fung_otu_sums <- rownames_to_column(fung_otu_sums, "ID")
fung_otu_sums <- arrange(fung_otu_sums, desc(sample_freq))
fung_otu_sums <- column_to_rownames(fung_otu_sums, "ID")

fung_otus_no_singlesamps <- fung_otu_sums[fung_otu_sums$sample_freq>1,]

fung_abun_no_singles <- fung_abun[rownames(fung_abun) %in% rownames(fung_otus_no_singlesamps),] # 18288 otus remain 
fung_abun_no_singles <- fung_abun_no_singles[,colSums(fung_abun_no_singles)>0] # 1424 samples
fung_met_no_singles <- fung_meta[as.character(fung_meta$x_seq_id) %in% names(fung_abun_no_singles),]
```


```{r}
new_fung_abun <- fung_abun_no_singles
new_fung_meta <- fung_met_no_singles

habs <- c("Terrestrial","Riverine", "Marine")
new_fung_meta <- new_fung_meta[which(new_fung_meta$habitat %in% habs),]

hab_subset <- function(otu_table, meta_table) {
  habs <- unique(meta_table$habitat)
  
  hab_list <- list()
  
  for (a_hab in habs) {
    sub_met <- meta_table[which(meta_table$habitat==a_hab),]
    
    sub_hab_abun <- otu_table[,which(names(otu_table) %in% sub_met$x_seq_id)]
    sub_hab_abun <- data.frame(sub_hab_abun)
    sub_hab_abun <- sub_hab_abun[which(rowSums(sub_hab_abun) > 0),]
    sub_hab_abun <- data.frame(sub_hab_abun)
    
    hab_list[[a_hab]] <- sub_hab_abun
    
  }
  
  return(hab_list)
}

fung_otus_by_hab <- hab_subset(new_fung_abun, new_fung_meta)



make_emp_hab_list <- function(hab, hab_list, meta_table, otu_table) {
  
  dat <- hab_list[[hab]]
  samples <- names(dat)
  
  sub_met <- meta_table[which(meta_table$x_seq_id %in% samples),]
  
  empo3 <- unique(sub_met$empo_3)
  emp_list <- list()
  
  for (e in empo3) {
    sub_sub_met <- sub_met[which(sub_met$empo_3==e),]
    sub_abun <- otu_table[,which(names(otu_table) %in% sub_sub_met$x_seq_id)]
    sub_abun <- data.frame(sub_abun)
    sub_abun <- sub_abun[which(rowSums(sub_abun) > 0),]
    sub_abun <- data.frame(sub_abun)
    
    otu_sums <- data.frame(rowSums(sub_abun))
    otu_sums$sample_freq <- apply(sub_abun, 1, function(c)sum(c!=0))
    names(otu_sums) <- c("otu_abundance", "sample_freq")
    otu_sums$tot_samps <- ncol(sub_abun)
    otu_sums$prev <- otu_sums$sample_freq/ncol(sub_abun)
    
    otu_sums <- rownames_to_column(otu_sums, "ID")
    otu_sums <- arrange(otu_sums, desc(sample_freq))
    otu_sums <- column_to_rownames(otu_sums, "ID")
    
    otu_sums <- otu_sums[otu_sums$prev>=0.2,] # set prevalence level here
    
    
    emp_list[[e]] <- otu_sums
  }
  
  return(emp_list)
}

#sites <- names(fung_otus_by_site)

fung_emp_otu_sums_by_hab <- lapply(habs,
                            make_emp_hab_list,
                            fung_otus_by_hab,
                            new_fung_meta,
                            new_fung_abun)

names(fung_emp_otu_sums_by_hab) <- names(fung_otus_by_hab)


all_things_listed <- lapply(fung_emp_otu_sums_by_hab, function(x) lapply(x, rownames))
flattened_ish <- lapply(all_things_listed, function(x) Reduce(c, x))

all_things <- unlist(flattened_ish)

all_things_unique <- unique(all_things)


new_fung_abun_prevfilter_on_empo <- new_fung_abun[rownames(new_fung_abun) %in% all_things_unique,]
new_fung_abun_prevfilter_on_empo <- new_fung_abun_prevfilter_on_empo[,colSums(new_fung_abun_prevfilter_on_empo)>0] #2128 OTUs, 1415 samples at 20% prevalence



test_abun <- new_fung_abun_prevfilter_on_empo
test_abun <- test_abun[rownames(test_abun) %in% which_ones,]
test_abun <- test_abun[,colSums(test_abun)>0]

# want to include a dummy row that includes the sum of reads for all removed OTUs
# spiec-easi does a total sum normalization, for which the total read count per sample must remain

# need to know what the colSums are for removed OTUs
fung_removed <- new_fung_abun[!rownames(new_fung_abun) %in% all_things_unique,]
fung_removed <- fung_removed[,names(fung_removed) %in% names(new_fung_abun_prevfilter_on_empo)]
fung_abun_with_dummy_row <- rbind(new_fung_abun_prevfilter_on_empo,
                                  colSums(fung_removed))
rownames(fung_abun_with_dummy_row)[nrow(fung_abun_with_dummy_row)] <- "dummy"


saveRDS(fung_abun_with_dummy_row, "../intermediates/global/prev20_fung_otu_table_with_dummy.rds")

filt_fung_meta <- new_fung_meta[new_fung_meta$x_seq_id %in% names(new_fung_abun_prevfilter_on_empo),]

```


## Bacteria


```{r}

bact_abun <- readRDS("../intermediates/global/bact_culled_abun_otu_lauras_method_post_matchup.rds")
bact_meta <- readRDS("../intermediates/global/bact_culled_meta_otu_lauras_method_post_matchup.rds")

# sample names in abundance table eventually get "X" appended due to the colnames being numeric
# make a new column with these names
#bact_meta$x_seq_ids <- paste0("X", bact_meta$sequencing_id)


bact_otu_sums <- data.frame(rowSums(bact_abun))
bact_otu_sums$sample_freq <- apply(bact_abun, 1, function(c)sum(c!=0))
names(bact_otu_sums) <- c("otu_abundance", "sample_freq")

bact_otu_sums <- rownames_to_column(bact_otu_sums, "ID")
bact_otu_sums <- arrange(bact_otu_sums, desc(sample_freq))
bact_otu_sums <- column_to_rownames(bact_otu_sums, "ID")

bact_otus_no_singlesamps <- bact_otu_sums[bact_otu_sums$sample_freq>1,]

bact_abun_no_singles <- bact_abun[rownames(bact_abun) %in% rownames(bact_otus_no_singlesamps),] # 77538 otus remain 
bact_abun_no_singles <- bact_abun_no_singles[,colSums(bact_abun_no_singles)>0] # 1434 samples
bact_met_no_singles <- bact_meta[as.character(bact_meta$x_seq_id) %in% names(bact_abun_no_singles),]
```


```{r}
new_bact_abun <- bact_abun_no_singles
new_bact_meta <- bact_met_no_singles


hab_subset <- function(otu_table, meta_table) {
  habs <- unique(meta_table$habitat)
  
  hab_list <- list()
  
  for (a_hab in habs) {
    sub_met <- meta_table[which(meta_table$habitat==a_hab),]
    
    sub_hab_abun <- otu_table[,which(names(otu_table) %in% sub_met$x_seq_id)]
    sub_hab_abun <- data.frame(sub_hab_abun)
    sub_hab_abun <- sub_hab_abun[which(rowSums(sub_hab_abun) > 0),]
    sub_hab_abun <- data.frame(sub_hab_abun)
    
    hab_list[[a_hab]] <- sub_hab_abun
    
  }
  
  return(hab_list)
}

bact_otus_by_hab <- hab_subset(new_bact_abun, new_bact_meta)



make_emp_hab_list <- function(hab, hab_list, meta_table, otu_table) {
  
  dat <- hab_list[[hab]]
  samples <- names(dat)
  
  sub_met <- meta_table[which(meta_table$x_seq_id %in% samples),]
  
  empo3 <- unique(sub_met$empo_3)
  emp_list <- list()
  
  for (e in empo3) {
    sub_sub_met <- sub_met[which(sub_met$empo_3==e),]
    sub_abun <- otu_table[,which(names(otu_table) %in% sub_sub_met$x_seq_id)]
    sub_abun <- data.frame(sub_abun)
    sub_abun <- sub_abun[which(rowSums(sub_abun) > 0),]
    sub_abun <- data.frame(sub_abun)
    
    otu_sums <- data.frame(rowSums(sub_abun))
    otu_sums$sample_freq <- apply(sub_abun, 1, function(c)sum(c!=0))
    names(otu_sums) <- c("otu_abundance", "sample_freq")
    otu_sums$tot_samps <- ncol(sub_abun)
    otu_sums$prev <- otu_sums$sample_freq/ncol(sub_abun)
    
    otu_sums <- rownames_to_column(otu_sums, "ID")
    otu_sums <- arrange(otu_sums, desc(sample_freq))
    otu_sums <- column_to_rownames(otu_sums, "ID")
    
    otu_sums <- otu_sums[otu_sums$prev>=0.2,] # set prevalence value
    
    emp_list[[e]] <- otu_sums
  }
  
  return(emp_list)
}

#sites <- names(bact_otus_by_site)

bact_emp_otu_sums_by_hab <- lapply(habs,
                            make_emp_hab_list,
                            bact_otus_by_hab,
                            new_bact_meta,
                            new_bact_abun)

names(bact_emp_otu_sums_by_hab) <- names(bact_otus_by_hab)


all_things_listed <- lapply(bact_emp_otu_sums_by_hab, function(x) lapply(x, rownames))
flattened_ish <- lapply(all_things_listed, function(x) Reduce(c, x))

all_things <- unlist(flattened_ish)

all_things_unique <- unique(all_things)


bact_abun_prevfilter_on_empo <- bact_abun[rownames(bact_abun) %in% all_things_unique,]
bact_abun_prevfilter_on_empo <- bact_abun_prevfilter_on_empo[,colSums(bact_abun_prevfilter_on_empo)>0] #13469 OTUs, 1434 samples at 20% prevalence

# want to include a dummy row that includes the sum of reads for all removed OTUs
# spiec-easi does a total sum normalization, for which the total read count per sample must remain

# need to know what the colSums are for removed OTUs
bact_removed <- new_bact_abun[!rownames(new_bact_abun) %in% all_things_unique,]
bact_removed <- bact_removed[,names(bact_removed) %in% names(bact_abun_prevfilter_on_empo)]
bact_abun_with_dummy_row <- rbind(bact_abun_prevfilter_on_empo,
                                  colSums(bact_removed))
rownames(bact_abun_with_dummy_row)[nrow(bact_abun_with_dummy_row)] <- "dummy"


saveRDS(bact_abun_with_dummy_row, "../intermediates/global/prev20_bact_otu_table_with_dummy.rds")

filt_bact_meta <- new_bact_meta[which(as.character(new_bact_meta$x_seq_id) %in% names(bact_abun_prevfilter_on_empo)),]
```


# Match up the fully filtered fungi and bacterial tables

```{r}

filtered_fungi <- fung_abun_with_dummy_row  # 1416 samples

filtered_bact <- bact_abun_with_dummy_row # 1434 samples

fung_meta_match <- filt_fung_meta[filt_fung_meta$sample_id %in% filt_bact_meta$sample_id,]

bact_meta_match <- filt_bact_meta[filt_bact_meta$sample_id %in% filt_fung_meta$sample_id,]

matched_ids <- fung_meta_match$sample_id # 1384 samples


# subset otu tables by samples matched in 16S and ITS
fung_otu_match <- fung_abun_with_dummy_row[,colnames(fung_abun_with_dummy_row) %in% fung_meta_match$x_seq_id] # 1384 samples remain
fung_otu_match <- fung_otu_match[rowSums(fung_otu_match)>0,] # 2129 otus remain (0 removed)

bact_otu_match <- bact_abun_with_dummy_row[,colnames(bact_abun_with_dummy_row) %in% bact_meta_match$x_seq_id] # 1384 samples remain
bact_otu_match <- bact_otu_match[rowSums(bact_otu_match)>0,] # 13469 otus remain (0 removed)


saveRDS(fung_otu_match, "../intermediates/global/fully_filtered_p20_fungal_otu_table_matched_up.rds")
saveRDS(fung_meta_match, "../intermediates/global/fully_filtered_p20_fungal_otu_metadata_matched_up.rds")
saveRDS(bact_otu_match, "../intermediates/global/fully_filtered_p20_bact_otu_table_matched_up.rds")
saveRDS(bact_meta_match, "../intermediates/global/fully_filtered_p20_bact_otu_metadata_matched_up.rds")

```








